components/usage/pkg/db/workspace.go

```diff
@@ -59,15 +60,33 @@ const (
 	WorkspaceType_Regular  WorkspaceType = "regular"
 )
 
+const listBatchSize = 2 ^ 16 - 1
+
 func ListWorkspacesByID(ctx context.Context, conn *gorm.DB, ids []string) ([]Workspace, error) {
 	if len(ids) == 0 {
 		return nil, nil
 	}
 
 	var workspaces []Workspace
-	tx := conn.WithContext(ctx).Where(ids).Find(&workspaces)
-	if tx.Error != nil {
-		return nil, fmt.Errorf("failed to list workspaces by id: %w", tx.Error)
+
+	items := len(ids)
+	batches := int(math.Ceil(float64(items / listBatchSize)))
+	for i := 0; i <= batches; i++ {
```

Comment:
Won't this always process one more batch than necessary?

(`math.Ceil` already rounds up the number of necessary batches, then you go from `0` to `batches` included?)

```suggestion
	batches := int(math.Ceil(float64(items / listBatchSize)))
	for i := 0; i < batches; i++ {
```

Reply:
I've simplified this a bit. The issue was with `float64(items / listBatchSize)` which actually performed integer division (div). What it needs to be is `float64(items) / listBatchSize` because the numerator determines the resulting type of the division operation, here it's floating point division. Which then works with Ceil to always run at least one batch.